### ISSUE TITLE
Alarm templates for all dimensions of a chart

### ISSUE THREAD

**Issue by Sakvar** [netdata/netdata#2107]
I made a python plugin which shows one chart with 7 dimensions.
My health config file is:
```
alarm: tempeature_out
on: ahu_local.temp
lookup: max -10m every 1m of dim1,dim2,dim3,dim4
units: C
every: 1m
warn: $this>20
delay: 1m
info: temperature is out of spec 
to: sysadmin
```
Alarm is shown up for a max of the sum of dimensions which are in `lookup` option.
How can I set a separate alarms for each dimension?

**Comment by l2isbad** [comment]
alarm: dim1_tempeature_out
on: ahu_local.temp
lookup: max -10m every 1m of dim1
units: C
every: 1m
warn: $this>20
delay: 1m
info: dim1 temperature is out of spec 
to: sysadmin

alarm: dim2_tempeature_out
on: ahu_local.temp
lookup: max -10m every 1m of dim2
units: C
every: 1m
warn: $this>20
delay: 1m
info: dim2 temperature is out of spec 
to: sysadmin

alarm: dim3_tempeature_out
on: ahu_local.temp
lookup: max -10m every 1m of dim3
units: C
every: 1m
warn: $this>20
delay: 1m
info: dim3 temperature is out of spec 
to: sysadmin

.
.
.

**Comment by Sakvar** [comment]
Is there the way to do it with one alarm?
In case I have 20 dimensions I need write 20 different alarms, wright?

**Comment by l2isbad** [comment]
> Is there the way to do it with one block?

I can be wrong but no. Db lookup is always the sum of all (all or defined group) dimensions.

> In case I have 20 dimensions I need write 20 different alarms, right?

All alarms in one file.

**Comment by ktsaou** [comment]
hm... alarm templates for dimensions... let me think a bit about it...

**Comment by ktsaou** [comment]
@cakrit 	this is on our roadmap. Low priority though.

**Comment by Pixcell** [comment]
Hello,

Is there an ETA templating alarms to watch all dimensions of the same chart separately ?

Thank you,

**Comment by cakrit** [comment]
No ETA yet unfortunately. We're getting some more help for the work on the daemon, so we'll probably look into this in the next couple of months.

**Comment by Pixcell** [comment]
@cakrit No problem, thank you for letting me know !
I don't mind trying to give it a go though. Any hint or direction you could point me towards ?

**Comment by cakrit** [comment]
Awesome. From the investigation thus far, it looks like this:

Reuse existing functionality for alarm templates. 
On the "lookup" line, accept "foreach (simple pattern)", instead of the specification of dimensions. This should tell netdata to create a separate alarm entry (rrdcalc) for each matching dimension of the charts that match the template.

How to do it:
Alarms are created based on templates and linked to charts whenever a chart is added. See the call to `rrdcalctemplate_link_matching` in L735 of `database/rrdset.c`. 
You will see that `rrdcalctemplate_link_matching` in `database/rrdcalctemplate.c` currently calls `rrdcalc_create` only once. This is the first thing that needs to change. We'll need to iterate through the chart's dimensions, see if they match the simple pattern provided in the lookup's foreach, and call `rrdcalc_create` once for each matching dimension. Note that to achieve the desired effect, you'll have to pass a modified `RRDCALCTEMPLATE` to `rrdcalc_create`. Check also the alarm name, because we'll need to add the dimension name to the end of it, for each alarm we create.

The above will effectively create these alarms every time a chart is added and also every time the health configuration is reloaded (rrdcalctemplate_link_matching is also called from `health_reload_host` in `health/health.c`.

What we'll be missing, is to also create alarm entries for newly added dimensions. From what I can tell, what needs to change is `rrddim_add_custom` in `database/rrddim.c`, so that we again go through the process of checking if the new dimension matches an existing alarm template, so that a new alarm entry can be added.

**Comment by thiagoftsm** [comment]
Hi everyone,

I finished the PR #6560 now that brings the templates for all dimensions of a chart now, the team and our users are goingo to test it and when the tests finish it will be merged to bring what was requested here.

Best Regards!
